### ISSUE TITLE
Incorrect sharing between String literals allows improper modification

### ISSUE THREAD

**Issue by eregon** [jruby/jruby#8365]
See https://github.com/ruby/fiddle/pull/149#discussion_r1792608879
If you use that branch and remove the `omit` in `Fiddle::TestFunc#test_qsort1` it fails on the latest JRuby release like:
```
$ ruby -Ilib test/fiddle/test_func.rb

/home/eregon/.rubies/jruby-9.4.8.0/lib/ruby/gems/shared/gems/power_assert-2.0.1/lib/power_assert.rb:8: warning: tracing (e.g. set_trace_func) will not capture all events without --debug flag
/home/eregon/.rubies/jruby-9.4.8.0/lib/ruby/gems/shared/gems/power_assert-2.0.1/lib/power_assert.rb:8: warning: tracing (e.g. set_trace_func) will not capture all events without --debug flag
Fiddle::VERSION: 1.1.3
Loaded suite test/fiddle/test_func
Started
..F
==============================================================================================================================
Failure: test_qsort1(Fiddle::TestFunc)
test/fiddle/test_func.rb:94:in `block in test_qsort1'
     91:         assert_equal("1349", buff, bug4929)
     92: 
     93:         # Ensure the test didn't mutate String literals
  => 94:         assert_equal("93" + "41", untouched)
     95:       end
     96:     ensure
     97:       # We can't use ObjectSpace with JRuby.
/home/eregon/code/fiddle/lib/fiddle/closure.rb:20:in `create'
test/fiddle/test_func.rb:77:in `test_qsort1'
org/jruby/RubyKernel.java:1423:in `catch'
org/jruby/RubyKernel.java:1418:in `catch'
org/jruby/RubyKernel.java:1423:in `catch'
org/jruby/RubyKernel.java:1418:in `catch'
<"9341"> expected but was
<"1349">

diff:
? 9341
? 1  9
? ?  ?
```

Using `buff = +"9341" + ""` instead of `buff = +"9341"` for both occurrences seems to fix it.
But it seems clearly a bug because `+"9341"` should always have its own separate storage from `"9341"`.

**Comment by headius** [comment]
Is this a bug in the FFI fiddle or JRuby?

**Comment by eregon** [comment]
Of JRuby (otherwise I wouldn't file it here).
The test works fine on TruffleRuby when using the FFI fiddle.

**Comment by eregon** [comment]
I guess what happens is there is some copy-on-write for Strings, and JRuby copies from byte[] to native and then native to byte[] but fails to notice that byte[] is shared/copy-on-write and so should not be written to.

**Comment by headius** [comment]
Yes, we are probably not unsharing somewhere that the string is passed out to FFI, or else not marking it as shared so buffer gets copied when writing. It should be possible to demonstrate without fiddle, though.

**Comment by eregon** [comment]
> It should be possible to demonstrate without fiddle, though.

Yes, definitely should be reproducible with just FFI.
I don't have time to minimize the repro, it's already pretty small though.

**Comment by headius** [comment]
Working fix is in #8370. I don't like the special-cased logic here, which probably should go in the strategy for String pointers, but this is small and localized and works.

**Comment by headius** [comment]
This relates to the ffi/ffi#1080 discussion about whether String passed to native should propagate changes back. In this case, it's mutable and passed as a pointer, so you would expect to have changes propagate back based on that discussion. But if the String were frozen, I'm leaning toward it being an error.

As you said in [XXX](https://github.com/ffi/ffi/issues/1080#issuecomment-2029585052):

> It is indeed problematic if called with a frozen string. I guess either we wouldn't rb_str_modify() in that case and assume the C code won't mutate (but still dangerous) or we would do the rb_str_modify() always and so raise when passed a frozen string, which seems good as it seems risky to pass a frozen string to a C function

This issue is related to the "dangerous" case. The string being passed in this issue was not frozen, but it was shared and we need to `modify` to ensure it has its own copy of the string buffer. But in my patch I also have to check for a frozen string and `dup` it to preserve current FFI behavior. If the intent of the call was to get string data out of the function, a passed-in frozen string would appear to ignore any changes during the call.

Unfortunately making it error for any case where you pass a frozen string as a pointer means that you would have to create a mutable copy even if the target function would *not* mutate the incoming buffer.

As in the FFI issue, I'm not sure there's a good answer here. I wish it had the capabilities of jnr-ffi, where we can specify that a pointer argument is *either* "in" or "out" or both (the default behavior for jnr-ffi and the *only* behavior in Ruby FFI).
